src/Data/Text/Encoding.hs

```diff
@@ -187,32 +189,295 @@ decodeLatin1 bs = withBS bs $ \fp len -> runST $ do
 foreign import ccall unsafe "_hs_text_is_ascii" c_is_ascii
     :: Ptr Word8 -> Ptr Word8 -> IO CSize
 
-isValidBS :: ByteString -> Bool
+data Progression
+  = WriteAndAdvance Char Int
+  | NeedMore
+  | Invalid
+
+decodeChunks ::
+#if defined(ASSERTS)
+  HasCallStack =>
+#endif
+  (Bits w, Num w, Storable w)
+  => w
+  -> ( Int
+    -> ByteString
+    -> A.MArray s
+    -> Int
+    -> Int
+    -> Int
+    -> ST s (Maybe (
+        (A.MArray s -> Int -> Int -> Int -> ST s (DecodeResult Text ByteString w))
+        -> ST s (DecodeResult Text ByteString w))
```

Comment:
Isn't this just a Church-encoded tuple? Why not return a real tuple?

Reply:
`Just outerArgs -> outerArgs outer` is less verbose than `Just (dst', dstLen', srcOff', dstOff') -> outer dst' dstLen' srcOff' dstOff'`, and at the time I had the idea that `bulkDecodeF` would be called at multiple places (which turned out not to be the case).

Reply:
- It's less verbose but harder to understand.
- The type here is also not as precise as it should be, it contains more inhabitants than just Church encoded tuples, so it sends the reader on a goose chase to check that there are no additional tricks hiding there.
Maintenance-wise, a tuple will be less of a hassle.